lib/entice/logic/casting.ex

```diff
@@ -8,51 +8,131 @@ defmodule Entice.Logic.Casting do
     end
 
     def handle_call(event, entity), do: super(event, entity)
+
+    @doc "This event triggers when the cast ends, it resets the casting timer, calls the skill's callback, and triggers recharge_end after a while."
+    def handle_event({:cast_end, skill, cast_callback, recharge_callback}, entity) do
+      recharge_timer = recharge_start(skill.recharge_time, skill, recharge_callback)
+      after_cast_timer = after_cast_start(250) #Dunno how to define constants
+      cast_callback.(skill)
+      {:ok, entity |> update_attribute(Casting,
+        fn c ->
+          %Casting{c | casting_timer: nil,
+          after_cast_timer: after_cast_timer,
+          recharge_timers: c.recharge_timers |> Map.put(skill, recharge_timer)}
+        end)}
+    end
+
+    @doc "This event triggers when a skill's recharge period ends, it resets the recharge timer for the skill."
+    def handle_event({:recharge_end, skill, recharge_callback}, entity) do
+      recharge_callback.(skill)
+      {:ok, entity |> update_attribute(Casting, fn c -> %Casting{c | recharge_timers: c.recharge_timers |> Map.delete(skill)} end)}
+    end
+
+    def handle_event({:after_cast_end}, entity) do
+      {:ok, entity |> update_attribute(Casting, fn c -> %Casting{c | after_cast_timer: nil} end)}
+    end
+
+    def handle_event({:test_message_1}, entity) do
+      {:ok, entity}
+    end
+
+    def terminate(_reason, entity),
+    do: {:ok, entity |> remove_attribute(Casting)}
+
+    defp reduce_mana(entity, new_mana) do
+      entity |> update_attribute(Energy, fn e -> %Energy{e | mana: new_mana} end)
+    end
+
+    #EVENT TRIGGERS
+
+    #Waits for #cast_time milliseconds then triggers the cast_end event.
+    defp cast_start(cast_time, skill, cast_callback, recharge_callback)  do #We take cast_time as arg rather than use skill.cast_time in case we want to modify it before
+      self |> Process.send_after({:cast_end, skill, cast_callback, recharge_callback}, cast_time)
+    end
+
+    #Waits for #recharge_time milliseconds then triggers the recharge_end event.
+    defp recharge_start(recharge_time, skill, recharge_callback) do
+      self |> Process.send_after({:recharge_end, skill, recharge_callback}, recharge_time)
+    end
+
+    defp after_cast_start(after_cast_time) do
+      self |> Process.send_after({:after_cast_end}, after_cast_time)
+    end
+
+    #CAST CONDITIONS
+    defp enough_energy?({:ok, skill}, dmana) when dmana > 0, do: {:ok, skill}
+    defp enough_energy?({:ok, _skill}, _dmana), do: {:error, :not_enough_energy}
+
+    defp not_recharging?({:ok, skill}, nil = _recharge_timer), do: {:ok, skill}
+    defp not_recharging?(_skill, _recharge_timer), do: {:error, :still_recharging}
+
+    defp not_casting?({:ok, skill}, nil = _casting_timer, nil = _after_cast_timer), do: {:ok, skill}
+    defp not_casting?(_skill, _casting_timer, _after_cast_timer), do: {:error, :still_casting}
+
+    defp can_cast?(skill,  %Entity{attributes: %{
```

Comment:
this is just concerning code structure, but personally i think its more clear if we put functions in and order like:
- 1) some function
- some sub-function that is used by 1
- another sub function used by 1
- 2) some other function
- a function used by 2)

etc.

tl;dr: can you move the helpers from above used by this below it?